Allow manual exposure detection

spheroid opened the pull request:
Let the user run the exposure detection manually if the previous check has not been run in the last 24 hours.

Commit:
Implement manual exposure detection

Commit:
Add the correct Finnish UI texts

Commit:
Make sure the publisher has a value

Replace share() with multicast() because the former wont publish the existing value to a new subscriber.

Commit:
Add manual exposure check to the exposures view

Commit:
Fix layout bugs

Commit:
Initial English & Swedish translations

Commit:
Avoid updating last checked date when update fails

Commit:
Add disabled state to exposure view

* Hook up the API status to the existing detection status Combine chain to determine when the exposure detection cannot be run
* Clean up the exposure view UI creation
* Add new icon for the disabled state

Commit:
Clean up

solita-pekko commented on Koronavilkku/Koronavilkku/UI/Components/RoundedButton.swift:
```diff
@@ -9,11 +9,36 @@ class RoundedButton: UIButton {
+                animation.toValue = CGFloat(Double.pi * 2.0)
+                animation.duration = 2
+                animation.repeatCount = .greatestFiniteMagnitude
+                imageView?.layer.add(animation, forKey: nil)
```
Doesn't matter in practice but `isLoading=true` seems to get called twice (the second time occurs right before `isLoading` is set to `false`) (or X times if the task takes X minutes to complete) and then the animation gets added unnecessarily for the second time (but all is good since *all* animations are removed in the end). Ignore didSet if value didn't change?

spheroid replied:
Good catch! Fixed.

solita-pekko commented on Koronavilkku/Koronavilkku/UI/Components/RoundedButton.swift:
```diff
@@ -9,11 +9,36 @@ class RoundedButton: UIButton {
+            if isLoading {
+                backgroundColor = UIColor.Greyscale.backdropGrey
+                setImage(UIImage.init(named: "refresh")?.withTintColor(UIColor.Greyscale.mediumGrey), for: .normal)
+                setTitle(nil, for: .normal)
```
accessibilityLabel could be updated for the button while it is displaying progress. Now VoiceOver reads it as "_refresh, mahdollisesti päivitä_".

spheroid replied:
Good point. Fixed with proper label.

solita-pekko commented on Koronavilkku/Koronavilkku/UI/Components/RoundedButton.swift:
```diff
@@ -9,11 +9,36 @@ class RoundedButton: UIButton {
+                imageView?.layer.removeAllAnimations()
+                setImage(nil, for: .normal)
+                setTitle(title, for: .normal)
+                self.setEnabled(true)
```
When VoiceOver is enabled and exposure check completed, then the button disappears but the voiceoverfocus remains on the button for a while before updating. Perhaps call `UIAccessibility.post(notification: .layoutChanged, argument: errorLabel)` etc. at this point to force an update (preferably the focus would move to the exposure information card's body label)?

spheroid replied:
Done.

solita-pekko commented on Koronavilkku/Koronavilkku/UI/Components/RoundedButton.swift:
```diff
@@ -9,11 +9,36 @@ class RoundedButton: UIButton {
+                imageView?.layer.removeAllAnimations()
+                setImage(nil, for: .normal)
+                setTitle(title, for: .normal)
+                self.setEnabled(true)
```
`setEnabled(false)` is not called when isLoading is set to true (for visual reasons perhaps?) - is `setEnabled(true)` needed here?

solita-pekko commented on Koronavilkku/Koronavilkku/View/Home/WideRowElement.swift:
```diff
@@ -67,114 +73,201 @@ final class ExposuresElement: WideRowElement {
-            make.bottom.left.right.equalToSuperview()
+            bodyView.snp.makeConstraints { make in
+                make.top.equalTo(titleView.snp.bottom).offset(10)
+                make.left.right.equalToSuperview()
```
I wonder if there is some autolayout magic spell that would allow aligning the body right edge either to the imageView's left edge or to card's left edge, depending on the titleView height: if the titleView height is >= imageView.height, then to card's left edge → would look nicer when a large font size is used (now when aligned to imageView's left edge a lot of space ends up not being used).

solita-pekko commented on Koronavilkku/Koronavilkku/Repository/ExposureRepository.swift:
```diff
@@ -14,19 +24,96 @@ protocol ExposureRepository {
+        
+        return Future { promise in
+            self.detectionRunning = true
+            self.detectionTask = BackgroundTaskForNotifications.execute { [weak self] success in
```
Is it possible to start a manual check while a scheduled background task is already running, or vice versa?

solita-pekko reviewed (commented):
Added a few tiny thoughts that could be checked.

Commit:
Move manual check to background task logic

To avoid background & foreground tasks running in parallel, control the background task in the same place.

Commit:
Clean up

* Prevent the same detection status from being published multiple times
* Avoid starting multiple loading animations
* Fix retain cycles
* Extract action logic to separate method

Commit:
Improve accessibility

* Move manual detection to a custom action
* Add accessible loading state to RoundedButton
* Make the exposures element accessibility properties a bit more sensible
* Avoid unnecessary exposures element redraws on launch

Commit:
Reset the button color explicitly

solita-pekko reviewed (approved):
Checked the accessibility related changes + tested a bit, works better now (especially now that I learnt how to use those custom actions :) ).

Commit:
Make sure the button loading animation is restarted

Commit:
Update CHANGELOG.md and remove debug print()

Commit:
Update CHANGELOG.md

Commit:
Update English translation, clean up

Commit:
Update translation
